### PR TITLE
[SALEOR-3088] Date & DateTime attributes

### PR DESCRIPTION
**Pull request by piotrgrundas**
I want to merge this change because it adds date & date time attributes handling.

[Dashboard PR](https://github.com/mirumee/saleor-dashboard/pull/1180)

# Impact

* [x] New migrations
* [x] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [x] Documentation needs to be updated

# Pull Request Checklist

* [x] Privileged queries and mutations are guarded by proper permission checks
* [x] Database queries are optimized and the number of queries is constant
* [x] Database migration files are up to date
* [x] The changes are tested
* [x] GraphQL schema and type definitions are up to date
* [x] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-3088 Datetime & date attributes](https://app.clickup.com/t/2549495/SALEOR-3088)

File: saleor/graphql/attribute/types.py [hunk 1]
```diff
@@ -94,6 +98,11 @@ def prepare_reference(attribute):
             .then(prepare_reference)
         )
 
+    @staticmethod
+    @traced_resolver
```
Comment (korycins) [inline saleor/graphql/attribute/types.py:102]: IMO, no need to add `traced_resolver` here, looks like simple resolver. @fowczarek  Can you confirm?

File: saleor/graphql/attribute/utils.py [hunk 2]
```diff
@@ -196,6 +199,34 @@ def _pre_save_boolean_values(
...
+    def _pre_save_date_values(
+        cls,
+        instance: T_INSTANCE,
+        attribute: attribute_models.Attribute,
+        attr_values: AttrValuesInput,
+    ):
+        value = cast(datetime, attr_values.date)
+        defaults = {
+            "date_time": value,
+            "name": f"{attribute.name}: {value.strftime('%d %b %Y')}",
```
Comment (korycins) [inline saleor/graphql/attribute/utils.py:212]: Why don't you use iso-8601 format here? (same for `_pre_save_date_time_values`)

File: saleor/graphql/attribute/utils.py [hunk 3]
```diff
@@ -671,6 +711,23 @@ def validate_standard_attributes_input(
...
+def validate_date_time_input(
+    attribute: "Attribute",
+    attr_values: "AttrValuesInput",
+    attribute_errors: T_ERROR_DICT,
+    variant_validation: bool,
+):
+    is_date = attribute.input_type == AttributeInputType.DATE
+    is_date_time = attribute.input_type == AttributeInputType.DATE_TIME
+
+    if attribute.value_required and (
```
Comment (korycins) [inline saleor/graphql/attribute/utils.py:723]: Can we simplify this `if`(to reduce amount of `and` and `or`)?
This is only a suggestion but imho the current format is hard to read.
you can assign `is_date and not attr_values.date` to variable and pass it to if condition
you can also do something like this
```
if not attribute.value_required:
    return
```

File: saleor/attribute/migrations/0014_auto_20210619_1836.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    operations = [
+        migrations.AddField(
+            model_name='attributevalue',
+            name='date_time',
+            field=models.DateTimeField(blank=True, null=True),
+        ),
+        migrations.AlterField(
+            model_name='attribute',
+            name='input_type',
+            field=models.CharField(choices=[('dropdown', 'Dropdown'), ('multiselect', 'Multi Select'), ('file', 'File'), ('reference', 'Reference'), ('numeric', 'Numeric'), ('rich-text', 'Rich Text'), ('boolean', 'Boolean'), ('date', 'Date'), ('date-time', 'Date Time')], default='dropdown', max_length=50),
```
Comment (fowczarek) [inline saleor/attribute/migrations/0014_auto_20210619_1836.py:21]: Missing black.

File: saleor/attribute/__init__.py [hunk 5]
```diff
@@ -20,23 +22,20 @@ class AttributeInputType:
         (NUMERIC, "Numeric"),
         (RICH_TEXT, "Rich Text"),
         (BOOLEAN, "Boolean"),
+        (DATE, "Date"),
+        (DATE_TIME, "Date Time"),
     ]
 
     # list of the input types that can be used in variant selection
-    ALLOWED_IN_VARIANT_SELECTION = [DROPDOWN, BOOLEAN]
+    ALLOWED_IN_VARIANT_SELECTION = [DROPDOWN, BOOLEAN, NUMERIC]
```
Comment (fowczarek) [inline saleor/attribute/__init__.py:30]: Do we have tests for this?

File: saleor/graphql/attribute/types.py (hunk 1 above)
Comment (fowczarek) [reply saleor/graphql/attribute/types.py:102]: Yep. We should use `traced_resolver` only for complex resolvers.

File: saleor/graphql/product/bulk_mutations/products.py [hunk 6]
```diff
@@ -161,12 +161,19 @@ class BulkAttributeValueInput(InputObjectType):
...
     values = graphene.List(
         graphene.String,
-        required=True,
+        required=False,
         description=(
             "The value or slug of an attribute to resolve. "
             "If the passed value is non-existent, it will be created."
         ),
     )
+    boolean = graphene.Boolean(
```
Comment (fowczarek) [inline saleor/graphql/product/bulk_mutations/products.py:170]: Maybe we should add `validate_one_of_args_is_in_mutation` to mutation which use this input?

**Review by fowczarek — APPROVED** [review]

**Comment by fowczarek** [comment]
Do we have any tests for date & date-time attributes?

File: saleor/graphql/product/bulk_mutations/products.py (hunk 6 above)
Comment (piotrgrundas) [reply saleor/graphql/product/bulk_mutations/products.py:182]: Validation is handled via `AttributeAssignmentMixin` like in `ProductVariantCreate`

**Comment by piotrgrundas** [comment]
@fowczarek in progress. I'll ping u when they will be added.

File: saleor/graphql/product/tests/test_variant.py [hunk 7]
```diff
@@ -1753,6 +1855,93 @@ def test_update_variant_with_rich_text_attribute(
     product_variant_updated.assert_called_once_with(product.variants.last())
 
 
+# @freeze_time(datetime(2020, 5, 5, 5, 5, 5, tzinfo=pytz.utc))
```
Comment (korycins) [inline saleor/graphql/product/tests/test_variant.py:1858]: Commented code

File: saleor/tests/fixtures.py [hunk 8]
```diff
@@ -795,6 +794,66 @@ def color_attribute(db):
...
+        name="Release date",
+        type=AttributeType.PRODUCT_TYPE,
+        input_type=AttributeInputType.DATE,
+        filterable_in_storefront=True,
+        filterable_in_dashboard=True,
+        available_in_grid=True,
+    )
+
+    value = datetime.datetime(2020, 10, 5, tzinfo=pytz.utc)
+    AttributeValue.objects.create(
```
Comment (korycins) [inline saleor/tests/fixtures.py:810]: Can you create them in bulk? The same for `date_time_attribute`. 
Both of them are fixture so would be great to have DB hits as low as possible

File: saleor/graphql/product/bulk_mutations/products.py [hunk 9]
```diff
@@ -173,12 +173,19 @@ class BulkAttributeValueInput(InputObjectType):
...
+        required=False,
+        description=(
+            "The boolean value of an attribute to resolve."
+            "If the passed value is non-existent, it will be created."
+        ),
+    )
```
Comment (fowczarek) [inline saleor/graphql/product/bulk_mutations/products.py:188]: I think those changes are not related to Date & DateTime attributes. We should add them in separate PR. 
If we add changes not related to PR we can easily miss that we should add more tests for this bugfix. Our QA can miss that changes and don't perform required tests.

File: saleor/graphql/attribute/utils.py [hunk 10]
```diff
@@ -188,16 +190,38 @@ def _pre_save_boolean_values(
...
+        boolean = bool(attr_values.boolean)
         value, _ = get_or_create(
             attribute=attribute,
-            slug=slugify(f"{attribute.id}_{attr_values.boolean}", allow_unicode=True),
+            slug=slugify(f"{attribute.id}_{boolean}", allow_unicode=True),
             defaults={
-                "name": f"{attribute.name}: {'Yes' if attr_values.boolean else 'No'}",
-                "boolean": attr_values.boolean,
+                "name": f"{attribute.name}: {'Yes' if boolean else 'No'}",
+                "boolean": boolean,
```
Comment (fowczarek) [inline saleor/graphql/attribute/utils.py:199]: I think those changes are not related to Date & DateTime attributes. We should add them in separate PR. 
If we add changes not related to PR we can easily miss that we should add more tests for this bugfix. Our QA can miss that changes and don't perform required tests.

File: saleor/graphql/product/bulk_mutations/products.py (hunk 9 above)
Comment (piotrgrundas) [reply saleor/graphql/product/bulk_mutations/products.py:188]: QA is aware of the addition of the variant creator for boolean values here. 
It was explicitly said it will be added here since Paweł was out while handling boolean attrs.

File: saleor/csv/utils/export.py [hunk 11]
```diff
@@ -60,6 +61,31 @@ def get_filename(model_name: str, file_type: str) -> str:
     )
 
 
+def parse_input(data: Any) -> Dict[str, Union[str, dict]]:
```
[the remark this answers is not in the document]
Comment (piotrgrundas) [reply saleor/csv/utils/export.py:64]: Yes sir! I just wanted to test it in test deployment.

**Review by fowczarek — APPROVED** [review]

File: saleor/csv/utils/export.py [hunk 12]
```diff
@@ -60,6 +61,32 @@ def get_filename(model_name: str, file_type: str) -> str:
...
+    """Parse input to correct data types, since scope coming from celery will be parsed to strings."""
+    if "attributes" in data:
+        serialized_attributes = []
+
+        for attr in data.get("attributes") or []:
```
Comment (korycins) [inline saleor/csv/utils/export.py:69]: Can't it be `data.get("attributes", []):` ?

File: saleor/graphql/page/tests/test_page.py [hunk 13]
```diff
@@ -760,6 +769,84 @@ def test_create_page_with_page_reference_attribute(
     assert page_type_page_reference_attribute.values.count() == values_count + 1
 
 
+@freeze_time(datetime(2020, 5, 5, 5, 5, 5, tzinfo=pytz.utc))
+def test_create_page_with_date_attribute(
```
Comment (korycins) [inline saleor/graphql/page/tests/test_page.py:773]: IMHO this should be split into two tests. You're testing here `date` attribute but also `datetime` attribute.

File: saleor/graphql/product/tests/test_variant.py [hunk 14]
```diff
@@ -1360,6 +1375,91 @@ def test_create_variant_with_rich_text_attribute(
     created_webhook_mock.assert_called_once_with(product.variants.last())
 
 
+@patch("saleor.plugins.manager.PluginsManager.product_variant_created")
+@freeze_time(datetime(2020, 5, 5, 5, 5, 5, tzinfo=pytz.utc))
+def test_create_variant_with_date_attribute(
```
Comment (korycins) [inline saleor/graphql/product/tests/test_variant.py:1380]: Shouldn't be two separate tests? You are checking date and datetime attribute here.

File: saleor/graphql/product/tests/test_variant.py (hunk 7 above)
Comment (korycins) [reply saleor/graphql/product/tests/test_variant.py:1858]: Still commented code :D

File: saleor/graphql/product/tests/test_variant.py [hunk 15]
```diff
@@ -1753,6 +1855,93 @@ def test_update_variant_with_rich_text_attribute(
     product_variant_updated.assert_called_once_with(product.variants.last())
 
 
+# @freeze_time(datetime(2020, 5, 5, 5, 5, 5, tzinfo=pytz.utc))
+@patch("saleor.plugins.manager.PluginsManager.product_variant_updated")
+def test_update_variant_with_date_attribute(
```
Comment (korycins) [inline saleor/graphql/product/tests/test_variant.py:1860]: Two separate tests?

File: saleor/csv/utils/export.py (hunk 12 above)
Comment (piotrgrundas) [reply saleor/csv/utils/export.py:69]: I would leave coz I'm not sure if there is a case whrere attributes might be none.

File: saleor/graphql/product/tests/test_variant.py (hunk 14 above)
Comment (piotrgrundas) [reply saleor/graphql/product/tests/test_variant.py:1380]: Well true but essentially it uses the same save method and same mode field. So I would say it doesnt bring anything.
But sure, I'll add it.

Comment (korycins) [reply saleor/graphql/product/tests/test_variant.py:1380]: I saw that some tests are split for DateTime and Date and some are not. I don't treat DateTime and Date as the same value so I believe we should have separate tests for them. Or tests with `pytest.parametrize`

File: saleor/graphql/product/tests/test_variant.py (hunk 7 above)
Comment (piotrgrundas) [reply saleor/graphql/product/tests/test_variant.py]: https://media0.giphy.com/media/SwyTq2jJxc9im6BYnN/giphy.gif?cid=790b76116de2390473b85e38392e137e7e38be53d3276768&rid=giphy.gif&ct=g

**Review by korycins — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Review by IKarbowiak — APPROVED** [review]
